### PR TITLE
netbox: allow to disable schema fetch

### LINKED ISSUE #4789: Schema-less netbox
Currently each netbox connection is a huge object which remembers all the spaces, indexes, and collations. For each space it creates `struct tuple_format` object, count of which is very limited. Collation count is huge - even if you don't use them, netbox still downloads 53! collations, including Chines, African, Arabic, etc.
Moreover, **all of this** is downloaded again every time, when schema version is changed on the server anyhow. This slows down connect and reconnect. It adds network hops, and makes netbox consume much more memory.
There are some cases, when schema is not needed at all. For example, when only calls and evals are used. This is the `vshard.router` case, and of any other proxy. Drop of the schema would allow to speed up connect/reconnect, not to re-download the whole schema on every change on the server, not to occupy `struct tuple_format` cells, consume less memory, when proxy makes many connections.

### PR DESCRIPTION
**Pull request by olegrok**
Before this patch each connection reload server schema if it
changed. However in some cases it's actually redundant e.g. user
performs only call and eval operations. This patch introduces
"fetch_schema" option (true by default to keep backward
compatibility) that allows to ignore schema changes on the server.
As result remote spaces are unavailable and on_schema_reload
trigger doesn't work.

@TarantoolBot document
Title: Document fetch_schema netbox option

Currently user could specify "fetch_schema" flag in netbox
connection to control schema fetching from remote instance.
In case if this option is false remote spaces will be unavailable
and on_schema_reload trigger won't work.

```
c = netbox.connect(uri, {fetch_schema = false})
c.space -- always will be nil
c:on_schema_reload(function() end) -- never will be triggered
```

Closes #4789

### CODE REVIEW

File: src/box/lua/net_box.c
Comment (locker) [inline src/box/lua/net_box.c:2011]: Please sync the comment to this function.

File: src/box/lua/net_box.c
Comment (locker) [inline src/box/lua/net_box.c:2568]: I'd like to explicitly check the previous state here, no matter if `fetch_schema` is set or not. Let's change to

```C
assert(transport->state == NETBOX_AUTH ||
       transport->state == NETBOX_FETCH_SCHEMA);
```

Also, please mind indentation: we indent by parenthesis, using tabs first and then spaces.

File: src/box/lua/net_box.c
Comment (locker) [inline src/box/lua/net_box.c:2578]: You don't need to check `fetch_schema` here: the function will go return the new `schema_version`, then go to `netbox_transport_fetch_schema`, which will return the new `schema_version` without doing anything, then go back here and continue the loop. That's how it works with graceful shutdown, for example. Let's leave it as is.

File: test/box-luatest/net_box_test.lua
Comment (locker) [inline test/box-luatest/net_box_test.lua:275]: Nit: IMO all potential artifacts created in a test case should be dropped in `after_test` callback so that if a test case fails for some reason, other test cases won't be affected. Let's drop the test spaces in the `after_test` callback, just in case:

```lua
g.after_test('test_schemaless', function()
    g.server:exec(function()
        if box.space.test1 then box.space.test1:drop() end
        if box.space.test2 then box.space.test2:drop() end
    end)
end
```

**Review by locker — CHANGES_REQUESTED** [review]

File: src/box/lua/net_box.c
Comment (olegrok) [reply src/box/lua/net_box.c]: After resolving comment below it's possible that state is active. I changed assert to
```
assert(transport->state == NETBOX_ACTIVE ||
       transport->state == NETBOX_AUTH ||
       transport->state == NETBOX_FETCH_SCHEMA);
```

Comment (olegrok) [reply src/box/lua/net_box.c]: However such approach triggers lua callback that is anyway is ignored because it's active->active state change

File: src/box/lua/net_box.c
Comment (olegrok) [reply src/box/lua/net_box.c]: Ok. Fixed.

File: test/box-luatest/net_box_test.lua
Comment (olegrok) [reply test/box-luatest/net_box_test.lua]: Agree. Added after_test fixture.

**Review by locker — CHANGES_REQUESTED** [review]
Please don't set the 'full-ci' label until you get all approvals or unless you specifically need to test your code on all architectures, because you expect failing tests. The rationale is to reduce CI load.

File: src/box/lua/net_box.c
Comment (locker) [reply src/box/lua/net_box.c]: Let's fix it like this instead:

```C
	if (transport->state != NETBOX_ACTIVE &&
	    transport->state != NETBOX_GRACEFUL_SHUTDOWN) {
		assert(transport->state == NETBOX_AUTH ||
		       transport->state == NETBOX_FETCH_SCHEMA);
```

Comment (olegrok) [reply src/box/lua/net_box.c]: Yes, it's better. Fixed.

File: changelogs/unreleased/gh-4789-net-box-fetch-schema.md
Comment (olegrok) [reply changelogs/unreleased/gh-4789-net-box-fetch-schema.md]: Fixed

**Review by locker — APPROVED** [review]
LGTM. Thanks!

**Comment by olegrok** [comment]
@Gerold103, please take a look

**Review by Gerold103 — APPROVED** [review]

File: src/box/lua/net_box.c
Comment (olegrok) [reply src/box/lua/net_box.c]: Thanks for review. Fixed.

**Pull request merged**
